### ISSUE TITLE
[no-unused-vars] Potential False Positive with rest destructuring

### ISSUE THREAD

**Issue by sk-** [eslint/eslint#13616]
One pattern to remove some properties from an object is to use destructuring as follows:

```js
const {p1, p2, ...rest} = data;
```

However in this case the rule `no-unused-vars` will complain about variables `p1` and `p2` being unused.

Note that just deleting the properties from the object won't work if one is using typescript and the properties are not optional.

**Tell us about your environment**

Node version: v14.6.0
npm version: v6.14.7
Local ESLint version: v7.7.0 (Currently used)
Global ESLint version: Not found
Parser: default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  "parserOptions": {
    // Required for certain syntax usages
    "ecmaVersion": 2018
  },
  "env": {
    "node": true,
    "es6": true
  },
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
const data = {foo: 1, bar: 2, baz: 3};
const {foo, ...rest} = data;
console.info(`Do something with ${rest}`);
```

```bash
npx eslint destructuring.js
```

**What did you expect to happen?**
No complaints.

**What actually happened? Please include the actual, raw output from ESLint.**
```
/test/destructuring.js
  2:8  error  'foo' is assigned a value but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

**Are you willing to submit a pull request to fix this bug?**
If it's easy enough to fix.

**Comment by mdjermanovic** [comment]
Hi @sk-, thanks for the issue!

I think this is correct default behavior of the `no-unused-vars` rule, because the `foo` variable isn't explicitly used.

The rule can be configured to allow this pattern, by setting the [ignoreRestSiblings](https://eslint.org/docs/rules/no-unused-vars#ignorerestsiblings) option to `true`.

```js
/* eslint no-unused-vars: ["error", { "ignoreRestSiblings": true }] */

const data = {foo: 1, bar: 2, baz: 3};
const {foo, ...rest} = data; // no errors
console.info(`Do something with ${rest}`);
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC12YXJzOiBbXCJlcnJvclwiLCB7IFwiaWdub3JlUmVzdFNpYmxpbmdzXCI6IHRydWUgfV0gKi9cblxuY29uc3QgZGF0YSA9IHtmb286IDEsIGJhcjogMiwgYmF6OiAzfTtcbmNvbnN0IHtmb28sIC4uLnJlc3R9ID0gZGF0YTsgLy8gbm8gZXJyb3JzXG5jb25zb2xlLmluZm8oYERvIHNvbWV0aGluZyB3aXRoICR7cmVzdH1gKTsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjksInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

**Comment by sk-** [comment]
Thanks @mdjermanovic, and sorry for not checking the rule's options before filing the issue.
